File: gadapter/src/main/java/com/zeyad/gadapter/Observables/ItemClickObservable.java
package com.zeyad.gadapter.Observables;

import com.zeyad.gadapter.GenericRecyclerViewAdapter;
import com.zeyad.gadapter.ItemInfo;

import io.reactivex.Observable;
import io.reactivex.Observer;
import io.reactivex.android.MainThreadDisposable;

import static com.zeyad.gadapter.Utils.checkMainThread;

public final class ItemClickObservable extends Observable<ClickEvent> {
    private final GenericRecyclerViewAdapter genericRecyclerViewAdapter;

    public ItemClickObservable(GenericRecyclerViewAdapter genericRecyclerViewAdapter) {
        this.genericRecyclerViewAdapter = genericRecyclerViewAdapter;
    }

    @Override
    protected void subscribeActual(Observer<? super ClickEvent> observer) {
        if (!checkMainThread(observer)) {
            return;
        }
        Listener listener = new Listener(observer);
        observer.onSubscribe(listener);
        genericRecyclerViewAdapter.setOnItemClickListener(listener.onItemClickListener);
    }

    final class Listener extends MainThreadDisposable {
        private final GenericRecyclerViewAdapter.OnItemClickListener onItemClickListener;

        Listener(final Observer<? super ClickEvent> observer) {
            this.onItemClickListener = new GenericRecyclerViewAdapter.OnItemClickListener() {
                @Override
                public void onItemClicked(int position, ItemInfo itemInfo, GenericRecyclerViewAdapter.ViewHolder holder) {
                    if (!isDisposed()) {
                        observer.onNext(new ClickEvent(position, itemInfo, holder));
                    }
                }
            };
        }

        @Override
        protected void onDispose() {
        }
    }
}

File: gadapter/src/main/java/com/zeyad/gadapter/observables/ItemClickObservable.kt
package com.zeyad.gadapter.observables

import com.zeyad.gadapter.GenericAdapter
import com.zeyad.gadapter.GenericViewHolder
import com.zeyad.gadapter.ItemInfo
import com.zeyad.gadapter.OnItemClickListener
import com.zeyad.gadapter.checkMainThread
import io.reactivex.Observable
import io.reactivex.Observer
import io.reactivex.android.MainThreadDisposable

class ItemClickObservable(private val genericRecyclerViewAdapter: GenericAdapter) : Observable<ClickEvent>() {

    override fun subscribeActual(observer: Observer<in ClickEvent>) {
        if (!observer.checkMainThread()) {
            return
        }
        val listener = Listener(observer)
        observer.onSubscribe(listener)
        genericRecyclerViewAdapter.onItemClickListener = listener.onItemClickListener
    }

    internal inner class Listener(observer: Observer<in ClickEvent>) : MainThreadDisposable() {
        internal val onItemClickListener: OnItemClickListener

        init {
            this.onItemClickListener = object : OnItemClickListener {
                override fun onItemClicked(position: Int, itemInfo: ItemInfo<*>, holder: GenericViewHolder<*>) {
                    if (!isDisposed) {
                        observer.onNext(ClickEvent(position, itemInfo, holder))
                    }
                }
            }
        }

        override fun onDispose() {}
    }
}
